File: spikectl/model/requests/hub.py
from __future__ import annotations

from base64 import b64encode
from spikectl.ujsonrpc import RPCRequest


class GetHubInfoRequest(RPCRequest):

    def __init__(self, idx: str = None):
        super().__init__('get_hub_info', {}, idx=idx)


class StartWriteProgramRequest(RPCRequest):

    def __init__(self, project_id: str, project_slot: int, name: str, project_type: str, size: str,
                 created: int, modified: int, idx: str = None):
        super().__init__('start_write_program', {
            'slotid': project_slot,
            'size': size,
            'meta': {
                'name': name,
                'project_id': project_id,
                'type': project_type,
                'created': created,
                'modified': modified
            }
        }, idx=idx)


class StartWriteResourceRequest(RPCRequest):

    def __init__(self, project_id: str, project_slot: int, name: str, project_type: str, size: str,
                 created: int, modified: int, idx: str = None):
        super().__init__('start_write_resource', {
            'slotid': project_slot,
            'size': size,
            'meta': {
                'name': name,
                'project_id': project_id,
                'type': project_type,
                'created': created,
                'modified': modified
            }
        }, idx=idx)


class GetFirmwareInfoRequest(RPCRequest):

    def __init__(self, idx: str):
        super().__init__('get_firmware_info', {}, idx=idx)


class TriggerCurrentStateRequest(RPCRequest):

    def __init__(self, idx: str=None):
        super().__init__('trigger_current_state', {}, idx=idx)


class SetHubNameRequest(RPCRequest):

    def __init__(self, name: str, idx: str = None):
        super().__init__('set_hub_name', {
            'name': b64encode(name)
        }, idx=idx)


class WritePackageRequest(RPCRequest):

    def __init__(self, data: str, transfer_id: int, idx: str = None):
        super().__init__('write_package', {
            'transferid': transfer_id,
            'data': data
        }, idx=idx)


class ProgramTerminateRequest(RPCRequest):

    def __init__(self, idx: str = None):
        super().__init__('program_terminate', {}, idx=idx)


class ProgramExecuteRequest(RPCRequest):

    def __init__(self, project_slot: int, idx: str = None):
        super().__init__('program_execute', {
            'slotid': project_slot
        }, idx=idx)


class MoveProjectRequest(RPCRequest):

    def __init__(self, old_slot: int, new_slot: int, idx: str = None):
        super().__init__('move_project', {
            'old_slotid': old_slot,
            'new_slotid': new_slot
        }, idx=idx)


class SwitchModeRequest(RPCRequest):

    def __init__(self, mode: str, idx: str = None):
        """

        :param idx:
        :param mode: 'play' or 'download'
        """

        super().__init__('program_modechange', {
            'mode': mode
        }, idx=idx)

    @classmethod
    def play(idx: str = None) -> SwitchModeRequest:
        return SwitchModeRequest('play', idx=idx)

    @classmethod
    def download(idx: str = None) -> SwitchModeRequest:
        return SwitchModeRequest('download', idx=idx)


class SyncDisplayRequest(RPCRequest):

    def __init__(self, sync: str, idx: str = None):
        super().__init__('sync_display', {
            'sync': sync
        }, idx=idx)


class ScratchResetYawRequest(RPCRequest):

    def __init__(self, sync: str, idx: str = None):
        super().__init__('scratch.reset_yaw', {}, idx=idx)

        
class ResetProgramTimeRequest(RPCRequest):

    def __init__(self, sync: str, idx: str = None):
        super().__init__('reset_program_time', {}, idx=idx)

class StartProgramTimeRequest(RPCRequest):

    def __init__(self, sync: str, idx: str = None):
        super().__init__('start_program_time', {}, idx=idx)

File: spikedev-js/src/requests/hub.js
import { RPCRequest } from '../ujsonrpc.js'

export class GetHubInfoRequest extends RPCRequest {
    constructor(idx) {
        super('get_hub_info', {}, idx)
    }
}

export class StartWriteProgramRequest extends RPCRequest {

    constructor(projectId, projectSlot, name, projectType, size, created, modified, idx = null) {
        super('start_write_program', {
            'slotid': projectSlot,
            'size': size,
            'meta': {
                'name': name,
                'project_id': projectId,
                'type': projectType,
                'created': created,
                'modified': modified
            }
        }, idx)
    }
}

export class StartWriteResourceRequest extends RPCRequest {
    constructor(project_id, project_slot, name, project_type, size, created, modified, idx = null) {
        super('start_write_resource', {
            'slotid': project_slot,
            'size': size,
            'meta': {
                'name': name,
                'project_id': project_id,
                'type': project_type,
                'created': created,
                'modified': modified
            }
        }, idx)
    }
}

export class GetFirmwareInfoRequest extends RPCRequest {
    constructor(idx = null) {
        super('get_firmware_info', {}, idx)
    }
}

export class TriggerCurrentStateRequest extends RPCRequest {

    constructor(idx = null) {
        super('trigger_current_state', {}, idx)
    }
}

export class SetHubNameRequest extends RPCRequest {
    constructor(name, idx = null) {
        super('set_hub_name', {
            'name': Buffer.from(name, 'ascii').toString('base64')
        }, idx)
    }
}

export class WritePackageRequest extends RPCRequest {
    constructor(data, transfer_id, idx = null) {
        super('write_package', {
            'transferid': transfer_id,
            'data': data
        }, idx)
    }
}

export class ProgramTerminateRequest extends RPCRequest {
    constructor(idx = null) {
        super('program_terminate', {}, idx)
    }
}

export class ProgramExecuteRequest extends RPCRequest {

    constructor(projectSlot, idx = null) {
        super('program_execute', {
            'slotid': projectSlot
        }, idx)
    }
}

export class MoveProjectRequest extends RPCRequest {

    constructor(oldSlot, newSlot, idx = null) {
        super('move_project', {
            'old_slotid': oldSlot,
            'new_slotid': newSlot
        }, idx)
    }
}

export class SwitchModeRequest extends RPCRequest {

    /**
     * 
     * @param {'play'|'download'} mode 
     * @param {string} [idx]
     */
    constructor(mode, idx = null) {
        super('program_execute', {
            'mode': mode
        }, idx)
    }

    static play(idx = null){
        return new SwitchModeRequest('play', idx)
    }

    static download(idx = null){
        return new SwitchModeRequest('download', idx)
    }
}

export class SyncDisplayRequest extends RPCRequest {
    /**
     * @param {string} sync
     * @param {string} [idx]
     */
    constructor(sync, idx = null) {
        super('sync_display', {
            'sync': sync
        }, idx)
    }
}

export class ScratchResetYawRequest extends RPCRequest {
    constructor(idx = null) {
        super('scratch.reset_yaw', {}, idx)
    }
}

export class ResetProgramTimeRequest extends RPCRequest {
    constructor(idx = null) {
        super('reset_program_time', {}, idx)
    }
}

export class StartProgramTimeRequest extends RPCRequest {
    constructor(idx = null) {
        super('start_program_time', {}, idx)
    }
}